src/backend/vulkan/src/xr.rs

```diff
@@ -0,0 +1,330 @@
+use ash::vk::{self, Handle};
+use hal::image;
+use std::{
+    ffi::{CStr, CString},
+    sync::{atomic::Ordering, Mutex},
+};
+
+#[cfg(feature = "use-openxr")]
+use once_cell::sync::Lazy;
+use std::sync::atomic::AtomicBool;
+
+use crate::{conv, native};
+
+#[cfg(feature = "use-openxr")]
+pub(crate) struct Instance {
+    instance: openxr::Instance,
+    system: openxr::SystemId,
+    required_device_extension_properties: Vec<CString>,
+    required_instance_extension_properties: Vec<CString>,
+    requirements: openxr::vulkan::Requirements,
+
+    session: Option<openxr::Session<openxr::Vulkan>>,
+    frame_waiter: Option<openxr::FrameWaiter>,
+    frame_stream: Option<openxr::FrameStream<openxr::Vulkan>>,
+    space: Option<openxr::Space>,
+    // FIXME add field for all -validations state, that'd track separate parts
```

Comment:
what kind of validation is supposed to be done here?

Reply:
It's a while since I wrote this part of code, but I'm somewhat sure I meant an `enum` that'd track the initialization state (`DeviceCreated`, `ReadyForRendering`, or something similar), that would reflect in the result of xr-related calls. For example, do not return handles if the creation process hasn't been finalized yet.